### ISSUE TITLE
INTERNAL_TOKEN_URI needlessly opened as RW

### ISSUE THREAD

**Issue by mapreri** [go-gitea/gitea#18655]
https://github.com/go-gitea/gitea/blob/7ce938b6c7a7c66eb615cf391e9742a65e3e3f95/modules/setting/setting.go#L1039

```go
func loadInternalToken(sec *ini.Section) string {
...
	switch tempURI.Scheme {
	case "file":
		fp, err := os.OpenFile(tempURI.RequestURI(), os.O_RDWR, 0600)
		if err != nil {
			log.Fatal("Failed to open InternalTokenURI (%s): %v", uri, err)
		}
...
```

Using `O_RDWR` is not really useful here.  I think that it should be possible to make it only `O_RDONLY`.

I spotted this as I tried to make the file `0400` and gitea failed to start.

Thanks for considering.

**Comment by wxiaoguang** [comment]
But the logic is: if the file doesn't exist, Gitea will try to generate a new secret and write it into the file.

**Comment by mapreri** [comment]
Right, so gitea can try to open the file (RO), if ENOENT then create with
the content (and close it, it's not even needed to read the file then
anyway).

Imho if the file exists but doesn't contain the token gitea could just
error out (didn't go check if this marches the current behaviour), or could
try to re-open the file as RW as needed if that's the current behaviour for
"file present but empty" and you want to keep it.


(I'm not proposing a patch because I really have no expirience with go
(which is something I should fix..), and also because I don't have time to
try figure out how to build the project and test such change. )

**Comment by mapreri** [comment]
Thank you!!

**Comment by mmoya** [comment]
Can #18657 be backported to 1.6? 

gitea 1.6.5 fails to start in kubernetes when `INTERNAL_TOKEN_URI` is pointing to a file mounted from a readonly secret. The error is  `...s/setting/setting.go:1085:loadInternalToken() [F] Failed to open InternalTokenURI (file:/etc/gitea/secrets/internal-token): open /etc/gitea/secrets/internal-token: permission denied`

**Comment by lunny** [comment]
> Can #18657 be backported to 1.6?
> 
> gitea 1.6.5 fails to start in kubernetes when `INTERNAL_TOKEN_URI` is pointing to a file mounted from a readonly secret. The error is `...s/setting/setting.go:1085:loadInternalToken() [F] Failed to open InternalTokenURI (file:/etc/gitea/secrets/internal-token): open /etc/gitea/secrets/internal-token: permission denied`

Will be released in v1.16.6
